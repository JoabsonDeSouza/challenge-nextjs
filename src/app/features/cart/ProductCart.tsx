import { Product } from '@/domain/products/product.model';
import { formatValueByCurrency } from '@/utils/functions';

interface ProductCartProps {
  product: Product;
}

export default function ProductCart({ product }: ProductCartProps) {
  return (
    <div
      key={product.id}
      className="bg-[#d4bd8d] rounded-lg shadow-md mb-4 flex"
    >
      <img
        src={product.image}
        alt={product.name}
        className="w-28 h-28 object-cover rounded-lg"
      />
      <div className="flex flex-col p-3 flex-1">
        <div className="flex flex-col flex-1">
          <div className="flex flex-col">
            <div className="flex gap-2">
              <p>Produto:</p>
              <p className="font-bold">{product.name}</p>
            </div>
            <div className="flex gap-2">
              <p>Valor:</p>
              <p className="font-bold">
                {formatValueByCurrency(product.price)}
              </p>
            </div>
            <div className="flex gap-2">
              <p>Quantidade:</p>
              <p className="font-bold">{product.quantity}</p>
            </div>
          </div>

          <span>{product.description}</span>
        </div>
      </div>
    </div>
  );
}
