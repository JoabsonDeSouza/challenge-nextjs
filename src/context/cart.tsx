import { Product } from '@/domain/products/product.model';
import { createContext, useContext, useState } from 'react';

type CartItem = Product & { quantity: number };

type CartContextType = {
  cart: CartItem[];
  updateCart: (item: CartItem) => void;
  clearCart: () => void;
};

const CartContext = createContext<CartContextType | undefined>(undefined);

export function CartProvider({ children }: { children: React.ReactNode }) {
  const [cart, setCart] = useState<CartItem[]>([]);

  const updateCart = (item: CartItem) => {
    setCart((prevCart) => {
      const existing = prevCart.find((p) => p.id === item.id);
      if (existing) {
        return prevCart.map((p) =>
          p.id === item.id ? { ...p, quantity: p.quantity + item.quantity } : p,
        );
      } else {
        return [...prevCart, item];
      }
    });
  };

  const clearCart = () => {
    setCart([]);
  };

  return (
    <CartContext.Provider value={{ cart, updateCart, clearCart }}>
      {children}
    </CartContext.Provider>
  );
}

export function useCartContext() {
  const context = useContext(CartContext);
  if (!context) {
    throw new Error('useCartContext deve ser usado dentro de um CartProvider');
  }
  return context;
}
