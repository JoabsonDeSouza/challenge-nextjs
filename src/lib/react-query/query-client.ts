import { QueryClient } from '@tanstack/react-query';

const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      refetchOnWindowFocus: false,
      staleTime: 1000 * 60 * 1,
      gcTime: 1000 * 60 * 60 * 1,
    },
  },
});

export { queryClient };
