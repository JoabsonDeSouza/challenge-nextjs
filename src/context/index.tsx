'use client';

import { AppReactProps } from '@/context/types';

import { CartProvider } from '@/context/cart';

const ContextsProvider = ({ children }: AppReactProps) => {
  return <CartProvider>{children}</CartProvider>;
};

export { ContextsProvider };
