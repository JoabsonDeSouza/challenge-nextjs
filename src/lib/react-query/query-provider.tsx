'use client';

import { queryClient } from '@/lib/react-query/query-client';
import { QueryClientProvider } from '@tanstack/react-query';
import { useEffect, useState } from 'react';

export const ReactQueryProvider = ({
  children,
}: {
  children: React.ReactNode;
}) => {
  const [isClient, setIsClient] = useState(false);

  useEffect(() => {
    setIsClient(true);
  }, []);

  if (!isClient) {
    return null;
  }

  return (
    <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
  );
};
