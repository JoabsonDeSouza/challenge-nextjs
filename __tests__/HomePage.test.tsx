import HomePage from '@/app/page';
import { CartProvider } from '@/context/cart';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { render } from '@testing-library/react';

const mockedProducts = [
  {
    name: 'Espresso',
    description: 'CafÃ© puro e forte em dose pequena',
    price: 822,
    image:
      'https://s2-g1.glbimg.com/detGX6jAIpR0gJnvyYW2-o4GEmI=/0x0:1920x1280/984x0/smart/filters:strip_icc()/i.s3.glbimg.com/v1/AUTH_59edd422c0c84a879bd37670ae4f538a/internal_photos/bs/2018/W/u/Q79lelQCKAYh1GlAu3ag/coffe-1354786-1920.jpg',
    id: 'd4a45678-4444-4dd4-ef44-456789012004',
    createdAt: '2025-05-13T20:05:15.000Z',
    category: 'Bebida',
    quantity: 0,
  },
  {
    name: 'Macchiato',
    description: 'Espresso com uma pequena quantidade de espuma de leite',
    price: 914,
    image:
      'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcQAwa7w5Fc-XqHlCX4Fxc2D1deyEofvlVig4Q&s',
    id: 'e5b56789-5555-4ee5-fg55-567890123005',
    createdAt: '2025-05-13T20:05:20.000Z',
    category: 'Bebida',
    quantity: 0,
  },
];

// jest.mock('../src/app/features/products/services/products.service', () => {
//   console.log('MOCKED products.service loaded');
//   return {
//     listProducts: jest.fn().mockResolvedValue(mockedProducts),
//     createProduct: jest.fn(),
//   };
// });

let queryClient: QueryClient;

beforeEach(() => {
  queryClient = new QueryClient();
});

function renderWithProviders(children: React.ReactElement) {
  return render(
    <QueryClientProvider client={queryClient}>
      <CartProvider>{children}</CartProvider>
    </QueryClientProvider>,
  );
}

describe('Home Page Snapshot', () => {
  it('should match snapshot before data loads', () => {
    const { asFragment } = renderWithProviders(<HomePage />);
    expect(asFragment()).toMatchSnapshot();
  });

  //TODO: Incomplete test case
  // it('should match snapshot after data loads', async () => {
  //   jest
  //     .spyOn(productService, 'listProducts')
  //     .mockResolvedValue(mockedProducts);

  //   const { asFragment, getByTestId } = renderWithProviders(<HomePage />);

  //   await waitFor(() => {
  //     expect(getByTestId('product-list')).toBeTruthy();
  //   });

  //   expect(asFragment()).toMatchSnapshot();
  // });
});
