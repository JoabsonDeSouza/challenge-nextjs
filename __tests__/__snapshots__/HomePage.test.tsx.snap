// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Page Snapshot should match snapshot before data loads 1`] = `
<DocumentFragment>
  <div
    class="container mx-auto justify-center flex items-center min-h-[calc(100vh-180px)]"
  >
    <p>
      Carregando...
    </p>
  </div>
</DocumentFragment>
`;
